shared-libs/rules-engine/src/task-fetcher.js

```diff
@@ -0,0 +1,210 @@
+      .then(docsToCommit => commitTaskDocs(db, docsToCommit))
+      .then(() => {
+        contactStateStore.markFresh(calculationTimestamp, dirtyContactIds);
+      });
+  };
+
+  if (contactIds) {
+    return refreshTasksForKnownContacts(db, calculationTimestamp, contactIds);
+  }
+
+  // If the contact state store does not contain all contacts, build up that list (contact doc ids + headless ids in reports/tasks)
+  if (!contactStateStore.hasAllContacts()) {
+    return refreshTasksForAllContacts(db, calculationTimestamp);
+  }
+
+  // Once the contact state store has all contacts, trust it and only refresh those marked dirty
+  return refreshTasksForKnownContacts(db, calculationTimestamp, contactStateStore.getContactIds());
+};
+
+const docsOf = query => query.then(result => result.rows.map(row => row.doc).filter(existing => existing));
+const fetch = {
+  allTaskData: db => {
+    const userDoc = contactStateStore.currentUser();
+    const userContactId = userDoc && userDoc._id;
+    return Promise.all([
+        docsOf(db.query('medic-client/contacts_by_type', { include_docs: true })),
+        docsOf(db.query('medic-client/reports_by_subject', { include_docs: true })),
+        fetch.allTasks(db, 'requester'),
+      ])
+      .then(([contactDocs, reportDocs, taskDocs]) => ({ contactDocs, reportDocs, taskDocs, userContactId }));
+  },
+
+  /*
+  PouchDB.query slows down when provided with a large keys array.
+  For users with ~1000 users it is ~50x faster to provider a start/end key instead of specifying all ids
+  */
+  allTasks: (db, prefix) => docsOf(db.query('medic-client/tasks', { startkey: `${prefix}-`, endkey: `${prefix}-\ufff0`, include_docs: true })),
+
+  existingContactStateStore: db => db.get(CONTACT_STATE_DOCID).catch(() => undefined),
+
+  tasksByRelation: (db, contactIds, prefix) => {
+    const keys = contactIds.map(contactId => `${prefix}-${contactId}`);
+    return docsOf(db.query('medic-client/tasks', { keys, include_docs: true }));
+  },
+
+  taskDataFor: (db, contactIds) => {
+    if (!contactIds || contactIds.length === 0) {
+      return Promise.resolve({});
+    }
+
+    const userDoc = contactStateStore.currentUser();
+    return docsOf(db.allDocs({ keys: contactIds, include_docs: true }))
+      .then(contactDocs => {
+        const subjectIds = contactDocs.reduce((agg, contactDoc) => {
+          registrationUtils.getSubjectIds(contactDoc).forEach(subjectId => agg.add(subjectId));
+          return agg;
+        }, new Set(contactIds));
+
+        return Promise.all([
+            docsOf(db.query('medic-client/reports_by_subject', { keys: Array.from(subjectIds).map(key => [key]), include_docs: true })),
+            fetch.tasksByRelation(db, contactIds, 'requester'),
+          ])
+          .then(([reportDocs, taskDocs]) => ({
+            userContactId: userDoc && userDoc._id,
+            contactDocs,
+            reportDocs,
+            taskDocs,
+          }));
+      });
+  },
+};
```

Comment:
Can you confirm you are proposing to move the `stateChangeCallback` function to a separate module?  Just wanting to clarify because I'm finding the line this comment is on to be slightly ambiguous.

Reply:
I was thinking the whole `const fetch = {...}` object. It appears to be a well encapsulated inner object that could easily be extracted. I may be missing something though...

Reply:
ahhh okay. that makes more sense

Reply:
i've completely decoupled task-fetcher from pouchdb